spec/abstract/Decode_A.thy

```diff
@@ -755,9 +757,9 @@ invocation is allowed.
 definition
   decode_invocation ::
   "bool \<Rightarrow> data \<Rightarrow> data list \<Rightarrow> cap_ref \<Rightarrow> cslot_ptr \<Rightarrow> cap \<Rightarrow>
-   (cap \<times> cslot_ptr) list \<Rightarrow> (invocation, 'z::state_ext) se_monad"
+   (cap \<times> cslot_ptr) list \<Rightarrow> obj_ref option \<Rightarrow> (invocation, 'z::state_ext) se_monad"
 where
-  "decode_invocation first_phase label args cap_index slot cap excaps \<equiv>
+  "decode_invocation first_phase label args cap_index slot cap excaps buffer \<equiv>
```

Comment:
Hmm. I am not sure where this change is coming from. Isn’t this `buffer` just the same thing as the third argument `args`?

Reply:
In other parts of the spec update it appears that `buffer` is replacing `args` but here this function still takes both. Is this how it’s supposed to be?

Reply:
Not unless there's something tricky going on. See args and buffer in `handle_invocation`
https://github.com/seL4/l4v/blob/rt/spec/abstract/Syscall_A.thy#L231

Reply:
See also in the source where buffer is passed through.
https://github.com/seL4/seL4/blob/master/src/api/syscall.c#L416

Reply:
Another question is if this is MCS specific or not, or, how this goes along with the case of master, as in, the argument `buffer` appears in the master version as well. How is this handled?

Reply:
in `handle_invocation`, `args` are read out of `buffer`. I don’t see why we need to pass both `args` and `buffer` to `decode_invocation`.

Reply:
I've asked Matt to have a look. I don't know this is supposed to work.

Reply:
Yes, I believe MCS does something new here. In particular, before MCS, there were no invocations other than endpoint IPC that needed to *write* to the IPC buffer. Pre-MCS invocation decodes only need to read the syscall args out of the IPC buffer. That meant that in the spec, we could read syscall args early in the decode, and pass those through to the more specific decode functions instead of the buffer. Endpoint IPC operations were special, in that they did (and still do) their own IPC buffer lookups, so we didn't need to pass the buffer through to endpoint IPC operations. The C doesn't have a nice way to pass a list of syscall args through, so the C passes the `buffer` pointer, and individual decode operations use it to call `getSyscallArgs`. In pre-MCS, I think this is all the `buffer` argument gets used for, so for that purpose it's equivalent to the spec-level `args` parameter.

Now, MCS has two new invocations that need to record consumed time in the IPC buffer. So now the (read-only) `args` are not enough. Those new invocations really need the `buffer` pointer. In the C, these invocations use the `buffer` pointer that was passed in.

So what do we do in the spec? I can see these options:
1. Do a second IPC buffer lookup in the invocation specs. But then in CRefine, we'd have to prove that this second lookup gives the same result as the first one in `handleInvocation`. That would be an annoying proof, and I think it would be confusing that the spec performs a lookup while the C doesn't.
2. Pass the `buffer` pointer through to all the invocations, and remove the `args` parameter from all the existing invocations, since we can get the `args` from the `buffer`. This would make the spec look like the C, but it would be a ton of rework for no particular benefit. Actually, I think it would make the spec worse, since it's no longer clear that for most invocations, all you need are the `args`. And the spec would no longer have a nicely factored out syscall args lookup.
3. Pass the `buffer` through only to the new invocations that need to write to it, and remove the `args` from those invocations only. This is a bit more reasonable that 2, but I think it's still annoying to have the syscall args lookup factored out for most invocations, but not for these two.
4. Pass the `buffer` and the `args` through to the new invocations, and leave everything else alone. This is what @MitchellBuckley has done. Although the `args` are redundant, I still think it's the nicest of these 4 options.

Reply:
Ok, thanks, Matt. I see. With this explanation, I do agree that 4 (what we have here) seems to be the best option for us. It will give us easier time merging with the master as well.

Reply:
> MCS has two new invocations that need to record consumed time in the IPC buffer, which need the IPC buffer pointer
>  the `buffer` and the `args` through to the invocations, although the `args` are redundant

Mitch, can you please update the commit message to something a bit more explanatory and include something like the above? Also maybe list what those invocations are?

Reply:
when you clean up later

Reply:
Ok, can do.

Reply:
Here's my suggestion for a commit message for the relevant commit:

```
rt aspec: correct IPC buffer pointer for sched context invocations

In the MCS kernel, two of the invocations on scheduling context
capabilities (consumed and yield-to) need to write the consumed time to
the current thread's IPC buffer.

Prior to this commit, ASpec was mistakenly using the first syscall
argument as the IPC buffer pointer for those invocations.

In this commit, we provide the correct IPC buffer pointer to those
invocations, by adding an extra `buffer` argument to the relevant
invocation decode functions. The IPC buffer pointer is obtained from the
existing call to `lookup_ipc_buffer` in `handle_invocation`. Syscall
arguments are dropped from decoding functions that no longer need them.

This change highlights differences between the pre-MCS and MCS kernels.
In the pre-MCS abstract specification, the only invocation that might
need to write to an IPC buffer is a IPC send on an endpoint. Endpoint
IPC performs its own IPC buffer lookup in `do_normal_transfer`. For all
other invocations, the IPC buffer was only needed for reading syscall
arguments. In the abstract specification, the syscall argument decoding
was therefore performed early in `handle_invocation`, and the syscall
args rather than the buffer were passed to the more specific decode
functions.

Now that MCS requires the IPC buffer pointer for some invocations, we
pass *both* the IPC buffer pointer and the syscall arguments to a small
number of the invocation decode functions, even though the syscall
arguments could be obtained from the IPC buffer. This allows us to keep
performing the syscall argument decoding early in `handle_invocation`,
and to maintain compatibility with existing proofs.

Note that in the C kernel, the IPC buffer pointer is always passed to
individual invocation decoding functions, and syscall arguments are only
decoded when they are needed.
```